File: leetcode/problems/problem-specific/5.longest-palindromic-substring/5.longest-palindromic-substring.go
package main

// * Two Pointer Solution | O(n^2) Time | O(1) Space

func longestPalindrome(s string) string {
	sSize := len(s)
	if sSize <= 1 {
		return s
	}

	res := ""
	for i := range s {
		// * Handle Odd length
		longestPalindromeHelper(s, i, i, &res)

		// * Handle Even length
		longestPalindromeHelper(s, i, i+1, &res)
	}

	return res
}

func longestPalindromeHelper(s string, left, right int, res *string) {
	sSize := len(s)
	for left >= 0 && right < sSize && s[left] == s[right] {
		curSize := right - left + 1
		if curSize > len(*res) {
			*res = s[left : right+1]
		}

		left--
		right++
	}
}

File: leetcode/problems/problem-specific/5.longest-palindromic-substring/5.longest-palindromic-substring.py
#
# @lc app=leetcode id=5 lang=python3
#
# [5] Longest Palindromic Substring
#

# @lc code=start

# * Two Pointers Solution | O(n^2) Time | O(1) Space
# * n -> The length of s string


class Solution:
    def longestPalindrome(self, s: str) -> str:
        longest_palindrome = ""

        def longestPalindromeHelper(left, right):
            nonlocal longest_palindrome

            # * Expand around its center, i.e., Start from the middle of the string and move the
            # * left pointer to left and right pointer to right to check if it's a palindrome.
            while left >= 0 and right < len(s) and s[left] == s[right]:
                cur_window_len = right - left + 1
                if cur_window_len > len(longest_palindrome):
                    longest_palindrome = s[left : right + 1]

                left -= 1
                right += 1

        for i in range(len(s)):
            # * Handle odd length palindromic string.
            longestPalindromeHelper(i, i)

            # * Handle even length palindromic string.
            longestPalindromeHelper(i, i + 1)

        return longest_palindrome


# @lc code=end
